Sources/PackageModel/Module.swift

```diff
@@ -119,19 +122,31 @@ extension ClangModule: XcodeModuleProtocol {
     }
 }
 
-public class TestModule: SwiftModule {
+let testModuleNameSuffix = "TestSuite"
+
+public protocol TestModuleProtocol {
     /// The base prefix for the test module, used to associate with the target it tests.
-    //
-    // FIXME: This seems like a fragile way to model this relationship.
-    public let basename: String
-    
-    public init(basename: String, sources: Sources) throws {
-        self.basename = basename
-        try super.init(name: basename + "TestSuite", sources: sources)
-        c99name = try PackageModel.c99name(name: basename) + "TestSuite"
+    var basename: String { get }
+    /// The name of the module.
+    var name: String { get }
+
+    associatedtype T
+    /// Helper method to create a test module instance.
+    static func testModule(basename: String, sources: Sources) throws -> T
+}
+
+extension TestModuleProtocol {
+    public var basename: String {
+        precondition(name.hasSuffix(testModuleNameSuffix))
+        return name[name.startIndex..<name.index(name.endIndex, offsetBy: -testModuleNameSuffix.characters.count)]
     }
 }
 
+extension SwiftModule: TestModuleProtocol {
+   public class func testModule(basename: String, sources: Sources) throws -> SwiftModule {
+        return try SwiftModule(name: basename + testModuleNameSuffix, isTest: true, sources: sources)
```

Comment:
This is actually a hack... I wanted to declare an init in the protocol but couldn't because SwiftModule isn't a final class.